Build failure: dmd

### Steps To Reproduce

Steps to reproduce the behavior:
1. build *dmd*

### Build log
The error occurs during `checkPhase`:
```
 ... fail_compilation/needspkgmod.d -verrors=0 -Icompilable  -fPIC ()==============================
Test 'fail_compilation/needspkgmod.d' failed. The logged output:
/build/dmd/generated/linux/release/64/dmd -conf= -m64 -Ifail_compilation -verrors=0 -Icompilable  -fPIC  -od/build/dmd/compiler/test/test_results/fail_compilation/d -of/build/dmd/compiler/test/test_results/fail_compilation/d/needspkgmod_0 -i=, fail_compilation/needspkgmod.d 
/nix/store/vcvhwiilizhijk7ywyn58p9l005n9sbn-binutils-2.43.1/bin/ld: /build/dmd/compiler/test/test_results/fail_compilation/d/needspkgmod_0.o: in function `D main':
fail_compilation/needspkgmod.d:(.text._Dmain[_Dmain]+0x5): undefined reference to `imports.pkgmod313.mod.bar()'
collect2: error: ld returned 1 exit status
Error: undefined reference to `imports.pkgmod313.mod.bar()`
       referenced from `_Dmain`
       perhaps a library needs to be added with the `-L` flag or `pragma(lib, ...)`
Error: linker exited with status 1
       g++ /build/dmd/compiler/test/test_results/fail_compilation/d/needspkgmod_0.o -o /build/dmd/compiler/test/test_results/fail_compilation/d/needspkgmod_0 -m64 -Xlinker -rpath=/build/dmd/compiler/test/../../../phobos/generated/linux/release/64 -L/build/dmd/compiler/test/../../../phobos/generated/linux/release/64 -lphobos2 -lpthread -lm -lrt -ldl 

==============================
Test 'fail_compilation/needspkgmod.d' failed: 
expected:
----
$r:.+_D7imports9pkgmod3133mod3barFZv.*Error: linker exited with status.+$
----
actual:
----
/nix/store/vcvhwiilizhijk7ywyn58p9l005n9sbn-binutils-2.43.1/bin/ld: {{RESULTS_DIR}}/fail_compilation/d/needspkgmod_0.o: in function `D main':
fail_compilation/needspkgmod.d:(.text._Dmain[_Dmain]+0x5): undefined reference to `imports.pkgmod313.mod.bar()'
collect2: error: ld returned 1 exit status
Error: undefined reference to `imports.pkgmod313.mod.bar()`
       referenced from `_Dmain`
       perhaps a library needs to be added with the `-L` flag or `pragma(lib, ...)`
Error: linker exited with status 1
       g++ {{RESULTS_DIR}}/fail_compilation/d/needspkgmod_0.o -o {{RESULTS_DIR}}/fail_compilation/d/needspkgmod_0 -m64 -Xlinker -rpath=/build/dmd/compiler/test/../../../phobos/generated/linux/release/64 -L/build/dmd/compiler/test/../../../phobos/generated/linux/release/64 -lphobos2 -lpthread -lm -lrt -ldl
----
diff:
----
-$r:.+_D7imports9pkgmod3133mod3barFZv.*Error: linker exited with status.+$
+/nix/store/vcvhwiilizhijk7ywyn58p9l005n9sbn-binutils-2.43.1/bin/ld: {{RESULTS_DIR}}/fail_compilation/d/needspkgmod_0.o: in function `D main':
+fail_compilation/needspkgmod.d:(.text._Dmain[_Dmain]+0x5): undefined reference to `imports.pkgmod313.mod.bar()'
+collect2: error: ld returned 1 exit status
+Error: undefined reference to `imports.pkgmod313.mod.bar()`
+       referenced from `_Dmain`
+       perhaps a library needs to be added with the `-L` flag or `pragma(lib, ...)`
+Error: linker exited with status 1
+       g++ {{RESULTS_DIR}}/fail_compilation/d/needspkgmod_0.o -o {{RESULTS_DIR}}/fail_compilation/d/needspkgmod_0 -m64 -Xlinker -rpath=/build/dmd/compiler/test/../../../phobos/generated/linux/release/64 -L/build/dmd/compiler/test/../../../phobos/generated/linux/release/64 -lphobos2 -lpthread -lm -lrt -ldl
----
```

Full log: https://hydra.nixos.org/build/275023230/nixlog/1

### Additional context
[Build failure on hydra](https://hydra.nixos.org/build/275023230)

### Notify maintainers
@lionello
@dukc
@jtbx

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/dm/dmd/generic.nix
+++ b/pkgs/by-name/dm/dmd/generic.nix
@@ -82,6 +82,13 @@ stdenv.mkDerivation (finalAttrs: {
       extraPrefix = "dmd/";
       hash = "sha256-N21mAPfaTo+zGCip4njejasraV5IsWVqlGR5eOdFZZE=";
     })
+  ] ++ lib.optionals (lib.versionOlder version "2.110.0") [
+    (fetchpatch {
+      url = "https://github.com/dlang/dmd/commit/fdd25893e0ac04893d6eba8652903d499b7b0dfc.patch";
+      stripLen = 1;
+      extraPrefix = "dmd/";
+      hash = "sha256-Uccb8rBPBLAEPWbOYWgdR5xN3wJoIkKKhLGu58IK1sM=";
+    })
   ];
 
   postPatch = ''

